### ISSUE TITLE
Frontend implementation of Number Mathesar type

### ISSUE THREAD

**Issue by kgodey** [centerofci/mathesar#258]
## Requirements

Users should be able to:
- view data in columns that have the Number Mathesar type
- change data type to and from the associated DB types based on options selected 
  - `NUMERIC`
  - `SMALLINT`
  - `INTEGER`
  - `BIGINT`
  - `DECIMAL`
  - `REAL`
  - `DOUBLE PRECISION`
- view, set, and change data type options related to the Number data types
- ~~filter data by Number-specific filtering options~~
- ~~group data by Number-specific grouping options~~
- use display options to view data in Number type columns
  - Numbers should be formatted according to the locale in the display option
  - If the percentage option is enabled, numbers should be shown as a percentage.
- see Number data shown in a monospace font or using tabular figures (see design spec)

Update:
- Grouping will be handled in #1073 
- Filtering will be handled in #1072 
 
## DB Type Mapping
- New columns of UI type Number and converting existing columns to UI type Number will use `NUMERIC`
- Applying database options will change DB type in the following ways:
    - Setting "Number Type" to `Integer` will update the DB type to `INTEGER` and will set "Integer data size" to 32 bits
        - Updating "Integer data size"  to 64 bits will change DB type to `BIGINT`
        - Updating "Integer data size"  to 16 bits will change DB type to `SMALLINT`
    - Setting "Number Type" to `Decimal` will leave the DB type as-is (it could be `NUMERIC` or `DECIMAL`) 
        - Setting decimal place and max digits will set the `precision` and `scale` type options. 
    - Setting "Number Type" to `Float` will change the DB type to `DOUBLE PRECISION` and will set "Floating Point Type" to "15 digits precision"
        - Updating "Floating Point Type"  to "6 digits precision" will change DB type to `REAL`
- Existing columns of the given types should automatically set the correct database options according to the logic above.
- `DECIMAL` columns should be treated the same as `NUMERIC` columns.

### Design
- The [Global Data Types](https://wiki.mathesar.org/design/specs/global-data-type-components) spec has the latest version of the UX.
- The [Number Data Type](https://wiki.mathesar.org/design/specs/number-data-type) spec has an older version of the UX but more information on user scenarios.

### Backend
- `/api/v0/tables/` lists the name and DB type of each columns 
- `/api/v0/tables/<id/columns/` lists column information in more detail
- `/api/v0/databases/<id>/types/` maps Mathesar types to DB types and provides a list of filtering and grouping options per-type.
  - These should be used to populate the filter and grouping options dynamically. 
- Updates to column types and type options should happen via `PATCH` requests to `/api/v0/tables/<id>/columns/<id>/`. These updates include
  - Changes to the data type (`type`)
  - Changes to the data type options (`type_options`)
  - Changes to the display options (`display_options`)

## Additional Context
- #238 (design)
- #269 (concept of Mathesar types)
- #498 (conversation about data types)
- Blocked by backend work:
  - #377
  - #378
  - #379
  - #380
  - #381
  - #382
  - #383
  - #386

**Comment by kgodey** [comment]
Marking this as ready since filtering and grouping have been moved to other issues and remaining backend work is done.

**Comment by seancolsen** [comment]
I'm closing this in favor of more specific tickets and outstanding PRs:

- #1340
- #1337
- #1339
- #1300
